### ISSUE TITLE
PartDesign: Simple Additive Loft consumes huge amount of resources due to mesh triangulation

### ISSUE THREAD

**Issue by GAZ082** [FreeCAD/FreeCAD#15755]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Sometimes i can, sometimes i cant finish this simple loft i'm attaching. Check the video (you may just skip to 2:0 or so).

Remove the .zip in the file.

[Loft Test.FCStd.zip](https://github.com/user-attachments/files/16488134/Loft.Test.FCStd.zip)

https://github.com/user-attachments/assets/d51c0796-1d97-4605-ba7f-e1881e303183

# System Details Report
---

## Report details
- **Date generated:**                              2024-08-04 12:50:06

## Hardware Information:
- **Hardware Model:**                              Gigabyte Technology Co., Ltd. P55-US3L
- **Memory:**                                      16,0 GiB
- **Processor:**                                   Intel® Core™ i5 760  × 4
- **Graphics:**                                    AMD Radeon™ RX 550 / 550 Series
- **Disk Capacity:**                               240,1 GB

## Software Information:
- **Firmware Version:**                            FD
- **OS Name:**                                     Arch Linux
- **OS Build:**                                    rolling
- **OS Type:**                                     64-bit
- **GNOME Version:**                               46
- **Windowing System:**                            Wayland
- **Kernel Version:**                              Linux 6.10.2-arch1-2

### Full version info

```shell
OS: Arch Linux (GNOME/gnome)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38385 (Git) AppImage
Build type: Release
Branch: main
Hash: 58241e027e7187a1ecaeeb75668ef9a497437c68
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: English/United States (en_US)
Installed mods: 
  * Curves.backup1713713759.0032098 0.6.29 (Disabled)
  * QuickMeasure 2022.10.28
  * ExtremeProDark 2.9.0
  * Curves.backup1693059780.2922277 0.6.13 (Disabled)
  * CfdOF.backup1710694906.6805208 1.24.9 (Disabled)
  * Assembly4.backup1696800837.713216 0.50.2 (Disabled)
  * Curves.backup1714093860.7275996 0.6.34 (Disabled)
  * OpenTheme 2024.5.16
  * CfdOF.backup1699129791.898633 1.24.7 (Disabled)
  * Assembly4.backup1710902627.637761 0.50.9 (Disabled)
  * Defeaturing.backup1710696629.0288966 1.2.2 (Disabled)
```

### Subproject(s) affected?

PartDesign

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by maxwxyz** [comment]
Could this be related to https://github.com/FreeCAD/FreeCAD/issues/15659 ?
Can you record a macro (start before beginning to reproduce) and end after the issue and attach it here?

**Comment by GAZ082** [comment]
@maxwxyz Now i managed to complete the loft (after about 2 minutes).
[Loft Test.FCStd.zip](https://github.com/user-attachments/files/16488632/Loft.Test.FCStd.zip)


Attaching macro and also file with the end result.
[Test2.FCMacro.zip](https://github.com/user-attachments/files/16488630/Test2.FCMacro.zip)

**Comment by maxwxyz** [comment]
@bgbsww could this be TNP related. And you have a macro to dive into.

**Comment by GAZ082** [comment]
@maxwxyz @bgbsww 

For a simple circle to square loft this seems to work, but the issue arises in my test, which is a circle and a slot.

This is also an issue in 0.21.2

**Comment by GAZ082** [comment]
Looks like this is working ok now. It still takes about 20 seconds in my (old) computer, but the ram consumption does skyrokets (about 3gb). I made some crazy lofts as test, all went fine.

![image](https://github.com/user-attachments/assets/ba0e9ff1-6892-4de3-a089-7b14a1ffc297)

I'll let you guys to decide on the closure of this ticket. Thanks!

@maxwxyz @bgbsww

**Comment by GAZ082** [comment]
![image](https://github.com/user-attachments/assets/4a491676-acd8-4dbb-8beb-9d50628aa5e4)
Mmmh, perhaps i should open another issue, but this is related i believe. After  clicking ok, the computer just keeps processing, has been like this for the past 2 minutes.

**Comment by GAZ082** [comment]
Created a separate issue for the pipe:

https://github.com/FreeCAD/FreeCAD/issues/15951

**Comment by AIRCAP** [comment]
fun fact. doing a **Part->Loft** between the 2 sketches is instant. it's just the **PardDesign->Additive loft** that takes forever.

**Comment by FEA-eng** [comment]
Still relevant

```
OS: Windows 10 build 19045
Architecture: x86_64
Version: 1.1.0dev.42564 (Git)
Build date: 2025/07/14 18:51:42
Build type: Release
Branch: main
Hash: dc2f2374b6ab3e3381653fad125b1b619bb49f7c
Python 3.11.13, Qt 6.8.3, Coin 4.0.3, Vtk 9.3.1, boost 1_86, Eigen3 3.4.0, PySide 6.8.3
shiboken 6.8.3, xerces-c 3.2.5, IfcOpenShell 0.8.2, OCC 7.8.1
Locale: Polish/Poland (pl_PL)
Navigation Style/Orbit Style/Rotation Mode: Blender/Trackball/Drag at cursor
Stylesheet/Theme/QtStyle: FreeCAD Light.qss/FreeCAD Light/
Logical DPI/Physical DPI/Pixel Ratio: 96/90.0633/1
```

**Comment by GAZ082** [comment]
@FEA-eng can you attach an example so I can test it as well? Thanks!

**Comment by FEA-eng** [comment]
> [@FEA-eng](https://github.com/FEA-eng) can you attach an example so I can test it as well? Thanks!

I tested it with the file you attached to this issue. And indeed a simple additive loft from a circle to a slot takes a lot of time to compute.

**Comment by GAZ082** [comment]
Interesting. This is still an issue in latest 1.1, but the thing is, the preview appears immediately but the control dialog doesnt. I had to close the program.  The memory does not goes up by much, but my 4 cores are used at their fullest.


https://github.com/user-attachments/assets/a1cde1a3-155f-4b0b-9b3e-aabdca2fa81c

**Comment by FEA-eng** [comment]
Still relevant

@maxwxyz It's a pretty nasty issue because it happens with a really basic operation. Do you know who could we ask to have a look at it?

```
OS: Windows 10 build 19045
Architecture: x86_64
Version: 1.2.0dev.20260120 (Git shallow)
Build date: 2026/01/20 21:54:07
Build type: Release
Branch: (HEAD detached at 9beca43)
Hash: 9beca4363ad327e453ee5a27f737537e31467901
Python 3.11.14, Qt 6.8.3, Coin 4.0.3, Vtk 9.3.1, boost 1_86, Eigen3 3.4.0, PySide 6.8.3
shiboken 6.8.3, xerces-c 3.3.0, IfcOpenShell 0.8.2, OCC 7.8.1
Locale: Polish/Poland (pl_PL)
Navigation Style/Orbit Style/Rotation Mode: Touchpad/Trackball/Drag at cursor
Stylesheet/Theme/QtStyle: FreeCAD.qss/FreeCAD Light/
Logical DPI/Physical DPI/Pixel Ratio: 96/113.268/1.25
```

**Comment by maxwxyz** [comment]
no idea. maybe @kadet1090 @Lgt2x or if it is occ @pieterhijma

**Comment by Lgt2x** [comment]
I confirm that the bottleneck is inside OCC, especially Delaunay triangulation operation. Not clear for now whether it's from bad OCC usage in FC, but we seem to create a lot of triangles.

<img width="1406" height="907" alt="Image" src="https://github.com/user-attachments/assets/25564fda-9dee-40c0-8e07-7b0bb2cfe163" />

**Comment by pieterhijma** [comment]
I don't think it is an OCC issue. Lofting on sketches is using breps, the rendering uses meshing. Additionally, the fact that lofting in Part is instant, is another indication that the problem is not OCC. My guess is that somehow rendering the shape is triggered too often or so in Part Design.

Part of the stacktrace of one such call into meshing is listed below:

<details>
<summary>stacktrace:</summary>
<pre>
#0  0x00007fff9d3a2c84 in BRepMesh_Delaun::createTrianglesOnNewVertices(NCollection_Shared<NCollection_DynamicArray<int>, void>&, Message_ProgressRange const&) () from /usr/lib/libTKMesh.so.7.9
#1  0x00007fff9d3a2f81 in BRepMesh_Delaun::AddVertices(NCollection_Shared<NCollection_DynamicArray<int>, void>&, Message_ProgressRange const&) () from /usr/lib/libTKMesh.so.7.9
#2  0x00007fff9d3caf08 in ?? () from /usr/lib/libTKMesh.so.7.9
#3  0x00007fff9d3ba718 in ?? () from /usr/lib/libTKMesh.so.7.9
#4  0x00007fff9d3b135d in ?? () from /usr/lib/libTKMesh.so.7.9
#5  0x00007fff9d3a3cc7 in BRepMesh_DelaunayBaseMeshAlgo::generateMesh(Message_ProgressRange const&) () from /usr/lib/libTKMesh.so.7.9
#6  0x00007fff9d38db6c in BRepMesh_BaseMeshAlgo::Perform(opencascade::handle<IMeshData_Face> const&, IMeshTools_Parameters const&, Message_ProgressRange const&) () from /usr/lib/libTKMesh.so.7.9
#7  0x00007fff9d3b8ff3 in ?? () from /usr/lib/libTKMesh.so.7.9
#8  0x00007fff9d3b17d7 in BRepMesh_FaceDiscret::process(int, Message_ProgressRange const&) const () from /usr/lib/libTKMesh.so.7.9
#9  0x00007fff9d3b1c6f in ?? () from /usr/lib/libTKMesh.so.7.9
#10 0x00007fff9b4ac1d7 in OSD_ThreadPool::performJob(opencascade::handle<Standard_Failure>&, OSD_ThreadPool::JobInterface*, int) () from /usr/lib/libTKernel.so.7.9
#11 0x00007fff9b4acbfb in OSD_ThreadPool::Launcher::run(OSD_ThreadPool::JobInterface&) () from /usr/lib/libTKernel.so.7.9
#12 0x00007fff9b4acc46 in OSD_ThreadPool::Launcher::perform(OSD_ThreadPool::JobInterface&) () from /usr/lib/libTKernel.so.7.9
#13 0x00007fff9d3b8403 in BRepMesh_FaceDiscret::performInternal(opencascade::handle<IMeshData_Model> const&, IMeshTools_Parameters const&, Message_ProgressRange const&) () from /usr/lib/libTKMesh.so.7.9
#14 0x00007fff9d37c1d9 in ?? () from /usr/lib/libTKMesh.so.7.9
#15 0x00007fff9d37d469 in IMeshTools_MeshBuilder::Perform(Message_ProgressRange const&) () from /usr/lib/libTKMesh.so.7.9
#16 0x00007fff9d3b87dc in BRepMesh_IncrementalMesh::Perform(opencascade::handle<IMeshTools_Context> const&, Message_ProgressRange const&) () from /usr/lib/libTKMesh.so.7.9
#17 0x00007fff9d3b8b89 in BRepMesh_IncrementalMesh::Perform(Message_ProgressRange const&) () from /usr/lib/libTKMesh.so.7.9
#18 0x00007fff9d3b8dde in BRepMesh_IncrementalMesh::BRepMesh_IncrementalMesh(TopoDS_Shape const&, IMeshTools_Parameters const&, Message_ProgressRange const&) () from /usr/lib/libTKMesh.so.7.9
#19 0x00007fff8aa44420 in PartGui::ViewProviderPartExt::setupCoinGeometry (shape=..., coords=0x55555ca04850, faceset=0x55555bdaead0, norm=0x55555ca04d70, lineset=0x55555bd901b0, nodeset=0x55555ca054e0, deviation=1, angularDeflection=1, normalsFromUV=true) at /home/pieter/current/fpa/freecad-main/src/Mod/Part/Gui/ViewProviderExt.cpp:1106
#20 0x00007fff8aa4662a in PartGui::ViewProviderPartExt::updateVisual (this=0x55555c9f8eb0) at /home/pieter/current/fpa/freecad-main/src/Mod/Part/Gui/ViewProviderExt.cpp:1492
#21 0x00007fff8aa43d40 in PartGui::ViewProviderPartExt::updateData (this=0x55555c9f8eb0, prop=0x55555d386f30) at /home/pieter/current/fpa/freecad-main/src/Mod/Part/Gui/ViewProviderExt.cpp:967
#22 0x00007fff8205a107 in PartDesignGui::ViewProviderBody::updateData (this=0x55555c9f8eb0, prop=0x55555d386f30) at /home/pieter/current/fpa/freecad-main/src/Mod/PartDesign/Gui/ViewProviderBody.cpp:256
#23 0x00007ffff5b2e47e in Gui::ViewProvider::update (this=0x55555c9f8eb0, prop=0x55555d386f30) at /home/pieter/current/fpa/freecad-main/src/Gui/ViewProvider.cpp:342
#24 0x00007ffff5b4ecc7 in Gui::ViewProviderDocumentObject::update (this=0x55555c9f8eb0, prop=0x55555d386f30) at /home/pieter/current/fpa/freecad-main/src/Gui/ViewProviderDocumentObject.cpp:419
#25 0x00007ffff517a973 in Gui::Document::slotChangedObject (this=0x5555561d01c0, Obj=..., Prop=...) at /home/pieter/current/fpa/freecad-main/src/Gui/Document.cpp:1118
</pre>
</details>

**Comment by FEA-eng** [comment]
> Additionally, the fact that lofting in Part is instant, is another indication that the problem is not OCC.

Right. Also, the preview (transparent or final result one) appears immediately. Only after confirming, the operation takes so long to finalize.

**Comment by luzpaz** [comment]
So the final verdict is that this is a Core issue not OCC or a combination of the 2 ?

**Comment by pieterhijma** [comment]
It's not an OCCT issue.

**Comment by Lgt2x** [comment]
needs testing against https://github.com/FreeCAD/FreeCAD/pull/32397
